icicle/appUtils/poseidon/poseidon.cuh

```diff
@@ -0,0 +1,141 @@
+#pragma once
+#include "../../curves/curve_config.cuh"
+#include "constants.cuh"
+
+#if !defined(__CUDA_ARCH__) && defined(DEBUG)
+#include <iomanip>
+#include <string>
+#include <sstream>
+#include <chrono>
+
+__host__ void print_scalar_t(scalar_t element) {
+  uint32_t * limbs = element.export_limbs();
+
+  std::stringstream hex_string;
+  hex_string << std::hex << std::setfill('0');
+
+  for (int i = 0; i < scalar_t::TLC; i++) {
+      hex_string << std::setw(8) << limbs[i];
+  }
+
+  std::cout << "0x" << hex_string.str() << std::endl;
+}
+
+__host__ void print_buffer_from_cuda(scalar_t * device_ptr, size_t size) {
+  scalar_t * buffer = static_cast< scalar_t * >(malloc(size * sizeof(scalar_t)));
+  cudaMemcpy(buffer, device_ptr, size * sizeof(scalar_t), cudaMemcpyDeviceToHost);
+
+  std::cout << "Start print" << std::endl;
+  for(int i = 0; i < size / 9; i++) {
+    std::cout << "State #" << i << std::endl;
+    for (int j = 0; j < 9; j++) {
+        print_scalar_t(buffer[i * 9 + j]);
+    }
+    std::cout << std::endl;
+  }
+  std::cout << std::endl;
+  free(buffer);
+}
+#endif
+
+#ifdef DEBUG
+__device__ void print_scalar(scalar_t element, int data) {
+    printf("D# %d, T# %d: 0x%08x%08x%08x%08x%08x%08x%08x%08x\n",
+           data,
+           threadIdx.x,
+           element.limbs_storage.limbs[0],
+           element.limbs_storage.limbs[1],
+           element.limbs_storage.limbs[2],
+           element.limbs_storage.limbs[3],
+           element.limbs_storage.limbs[4],
+           element.limbs_storage.limbs[5],
+           element.limbs_storage.limbs[6],
+           element.limbs_storage.limbs[7]
+    );
+}
+#endif
+
+struct PoseidonConfiguration {
+    uint partial_rounds, full_rounds_half, t;
+    scalar_t * round_constants, * mds_matrix, * non_sparse_matrix, *sparse_matrices;
+};
+
+class Poseidon {
+  public:
+    uint t;
+    PoseidonConfiguration config;
+
+    enum HashType {
+        ConstInputLen,
+        MerkleTree,
+    };
+
+    Poseidon(const uint arity) {
+        t = arity + 1;
+        this->config.t = t;
+
+        // Pre-calculate domain tags
+        // Domain tags will vary for different applications of Poseidon
+        uint32_t tree_domain_tag_value = 1;
+        tree_domain_tag_value = (tree_domain_tag_value << arity) - tree_domain_tag_value;
+        tree_domain_tag = scalar_t::from(tree_domain_tag_value);
+
+        const_input_no_pad_domain_tag = scalar_t::one();
+
+        // TO-DO: implement binary shifts for scalar type
```

Comment:
Maybe you can open an issue for this?

Reply:
Opened
https://github.com/ingonyama-zk/icicle/issues/76